src/Microsoft.DocAsCode.MarkdigEngine.Extensions/TripleColon/CodeExtension.cs

```diff
@@ -0,0 +1,258 @@
+            if(string.IsNullOrEmpty(code))
+            {
+                logError($"The code snippet \"{source}\" could not be found.");
+                return false;
+            }
+            
+            var updatedCode = GetCodeSnippet(range, id, code, logError).TrimEnd('\r').TrimEnd('\n');
+            updatedCodes.Add(updatedCode);
+
+            if (updatedCode == string.Empty)
+            {
+                return false;
+            }
+
+            if(string.IsNullOrEmpty(language))
+            {
+                language = InferLanguageFromFile(source, logError);
+            }
+
+            htmlAttributes = new HtmlAttributes();
+            htmlAttributes.AddProperty("class", $"lang-{language}");
+            if (!string.IsNullOrEmpty(interactive))
+            {
+                htmlAttributes.AddProperty("data-interactive", language);
+                htmlAttributes.AddProperty("data-interactive-mode", interactive);
+            }
+            if (!string.IsNullOrEmpty(highlight)) htmlAttributes.AddProperty("highlight-lines", highlight);
+
+            RenderDelegate = (renderer, obj) =>
+            {
+                renderer.WriteLine("<pre>");
+                renderer.Write("<code").WriteAttributes(obj).Write(">");
+                renderer.WriteLine(updatedCodes.FirstOrDefault());
+                updatedCodes.RemoveAt(0);
+                renderer.WriteLine("</code></pre>");
+
+                return true;
+            };
+
+            return true;
+        }
+
+        private string InferLanguageFromFile(string source, Action<string> logError)
+        {
+            var fileExtension = source.Split('.').ToList().LastOrDefault();
+            if(fileExtension == null)
+            {
+                logError($"Language is not set, and your source has no file type. Cannot infer language.");
+            }
+            var language = HtmlCodeSnippetRenderer.LanguageAlias.Where(oo => oo.Value.Contains(fileExtension) || oo.Value.Contains($".{fileExtension}")).FirstOrDefault();
+            return language.Key;
+        }
+
+        private static string GetCodeSnippet(string range, string id, string code, Action<string> logError)
+        {
+            if(!string.IsNullOrEmpty(range) && !string.IsNullOrEmpty(id))
+            {
+                logError($"You must set only either Range or Id, but not both.");
+            }
+            
+            var codeSections = new List<string>();
+
+            if (!string.IsNullOrEmpty(range))
+            {
+                codeSections = GetCodeSectionsFromRange(range, code, codeSections, logError);
+            }
+            else if (!string.IsNullOrEmpty(id))
+            {
+                var strRegexStart = @"((<|</|#region)\s*" + id + @"(>|(\s*>)))";
+                var idRegexStart = new Regex(strRegexStart, RegexOptions.IgnoreCase);
+                var idRegexEnd = new Regex(strRegexStart + @"|#endregion", RegexOptions.IgnoreCase);
```

Comment:
Creating a Regex is a relatively expensive operation, consider promoting it to a static variable so it is created only once. Static members usually starts with `s_` prefix.

```suggestion
                private static readonly Regex s_idRegexEnd = new Regex(strRegexStart + @"|#endregion", RegexOptions.IgnoreCase);
```

Reply:
Yeah, I thought about that, but if you look at line 155, you will see that I need to dynamically create it with the id value. Do you know of a better way to do that?

Reply:
Actually, I removed the need to use Regex :).